### PR TITLE
Added .gitattributes to manage end-of-line checks for Windows/*nix systems

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
There is a difference between end of line CRLF vs LF between Windows/*nix systems which causes `spotless` checks to fail. The suggested solution is to use `.gitattributes` , see please https://github.com/diffplug/spotless/issues/23

Linux:

```
$ ./gradlew spotlessJavaCheck
...
> Configure project :qa:os
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 7.3
  OS Info               : Linux 5.13.0-22-generic (amd64)
  JDK Version           : 11 (Ubuntu JDK)
  JAVA_HOME             : /usr/lib/jvm/java-11-openjdk-amd64
  Random Testing Seed   : 2A834AAD373C92B6
  In FIPS 140 mode      : false
=======================================

BUILD SUCCESSFUL in 2s
169 actionable tasks: 169 up-to-date
```

Windows:
```
$ ./gradlew.bat spotlessJavaCheck

> Configure project :plugins:repository-hdfs
hdfsFixture unsupported, please set HADOOP_HOME and put HADOOP_HOME\bin in PATH

> Configure project :qa:os
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 7.3
  OS Info               : Windows 10 10.0 (amd64)
  JDK Version           : 11 (Amazon Corretto JDK)
  JAVA_HOME             : C:\Program Files\Amazon Corretto\jdk11.0.11_9
  Random Testing Seed   : EA3DA7454331063
  In FIPS 140 mode      : false
=======================================

BUILD SUCCESSFUL in 7s
169 actionable tasks: 169 up-to-date
```
 
### Issues Resolved
Closes https://github.com/opensearch-project/OpenSearch/issues/1637
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by owaiskazi19** [comment]
@reta Thanks for the PR. Coming from https://github.com/opensearch-project/OpenSearch/issues/1637, can you help me understand what we get on Windows after running `./gradlew.bat :benchmarks:spotlessApply`? I'm just trying to understand if we need `.gitattributes`.

**Comment by dblock** [comment]
+1 on @owaiskazi19 question, we should have consistent CR/LFs in code regardless of whether someone edited a file on Windows or *nix

**Comment by reta** [comment]
@dblock @owaiskazi19 yes, sure:

```
$ ./gradlew.bat :benchmarks:spotlessApply                                                                             
                                                                                                                      
> Configure project :plugins:repository-hdfs                                                                          
hdfsFixture unsupported, please set HADOOP_HOME and put HADOOP_HOME\bin in PATH                                       
                                                                                                                      
> Configure project :qa:os                                                                                            
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.                               
=======================================                                                                               
OpenSearch Build Hamster says Hello!                                                                                  
  Gradle Version        : 7.3                                                                                         
  OS Info               : Windows 10 10.0 (amd64)                                                                     
  JDK Version           : 11 (Amazon Corretto JDK)                                                                    
  JAVA_HOME             : C:\Program Files\Amazon Corretto\jdk11.0.11_9                                               
  Random Testing Seed   : 89C568BCDCA3C3D8                                                                            
  In FIPS 140 mode      : false                                                                                       
=======================================                                                                               
                                                                                                                      
BUILD SUCCESSFUL in 6s                                                                                                
17 actionable tasks: 3 executed, 14 up-to-date                                                                        
```

Here is the diff:

```
$ git diff
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/fs/AvailableIndexFoldersBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/indices/breaker/MemoryStatsBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/routing/allocation/AllocationBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/routing/allocation/Allocators.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/search/aggregations/TermsReduceBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/search/aggregations/bucket/terms/LongKeyedBucketOrdsBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/search/aggregations/bucket/terms/StringTermsSerializationBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/time/DateFormatterBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/time/DateFormatterFromBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/benchmark/time/RoundingBenchmark.java.
The file will have its original line endings in your working directory
warning: CRLF will be replaced by LF in benchmarks/src/main/java/org/opensearch/common/RoundingBenchmark.java.
The file will have its original line endings in your working directory

```

**Comment by owaiskazi19** [comment]
Thanks for the output. Since `./gradlew.bat :benchmarks:spotlessApply` is successful `gradle check` on Windows should not break at least because of Spotless or is there anything else I'm missing here?

**Comment by reta** [comment]
> Thanks for the output. Since `./gradlew.bat :benchmarks:spotlessApply` is successful `gradle check` on Windows should not break at least because of Spotless or is there anything else I'm missing here

@owaiskazi19 I think you are mistaken `spotlessApply` for `spotlessJavaCheck`:
 `./gradlew.bat :benchmarks:spotlessApply` - does **reformat** the code (not part of the `check` or `build`)
 `./gradlew.bat :benchmarks:spotlessJavaCheck` - does **check the format**  of the code (part of the `check` or `build`)

The task which fails is `spotlessJavaCheck`, not `spotlessApply`

**Comment by owaiskazi19** [comment]
Sorry if I was not clear. I meant since we had a successful run of SpotlessApply, can't we just raise a PR for that formatted  change instead of adding `.gitattributes` file?

**Comment by reta** [comment]
> Sorry if I was not clear @reta. I meant since we had a successful run of SpotlessApply, can't we just raise a PR for that formatted change instead of adding `.gitattributes` file?

I suspect if we fix Windows formatting, than `spotlessJavaCheck` will fail on *nix :) The recipe with `.gitattributes` comes from `spotless` maintainers: https://github.com/diffplug/spotless/issues/39

**Comment by owaiskazi19** [comment]
> > Sorry if I was not clear @reta. I meant since we had a successful run of SpotlessApply, can't we just raise a PR for that formatted change instead of adding `.gitattributes` file?
> 
> I suspect if we fix Windows formatting, than `spotlessJavaCheck` will fail on *nix :) The recipe with `.gitattributes` comes from `spotless` maintainers: [diffplug/spotless#39](https://github.com/diffplug/spotless/issues/39)

I'm not sure if it will break on *nix for this specific case. Can you give it a try?
Also, if we plan to go with `.gitattributes`, don't we have to include `lineEndings 'GIT_ATTRIBUTES' ` as mentioned here:  https://github.com/diffplug/spotless/blob/9aefb65262685a9959360470b070e2e649f038eb/README.md#custom-rules. @dblock WDYT?

**Comment by reta** [comment]
> I'm not sure if it will break on *nix for this specific case. Can you give it a try?

I will, sure

> Also, if we plan to go with .gitattributes, don't we have to include lineEndings 'GIT_ATTRIBUTES' as mentioned here: https://github.com/diffplug/spotless/blob/9aefb65262685a9959360470b070e2e649f038eb/README.md#custom-rules. @dblock WDYT of .gitattributes?

Nope, it is a default for a long time: https://github.com/diffplug/spotless/issues/23

**Comment by reta** [comment]
@owaiskazi19 expectation marches, have a branch for you, https://github.com/reta/OpenSearch/tree/issue-1637.apply, what I did

On Windows:
- ./gradlew.bat :benchmarks:spotlessApply
- ./gradlew.bat :benchmarks:spotlessJavaCheck - SUCCESS
- commit & push

On Linux:
 - pull
 - ./gradlew.bat :benchmarks:spotlessJavaCheck - FAILURE

```
> Configure project :qa:os                                                
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.       
=======================================                                                                                                                                                      
OpenSearch Build Hamster says Hello!                                                          
  Gradle Version        : 7.3                                                                 
  OS Info               : Linux 5.13.0-22-generic (amd64)                                     
  JDK Version           : 11 (Ubuntu JDK)                                                     
  JAVA_HOME             : /usr/lib/jvm/java-11-openjdk-amd64                 
  Random Testing Seed   : BF2E36EEA16B5B41                                                    
  In FIPS 140 mode      : false                                                               
=======================================                                                       
                                                                                              
> Task :benchmarks:spotlessJavaCheck FAILED                                                   
                                                                                              
FAILURE: Build failed with an exception.   
```

File: .gitattributes
Comment (reta) [inline .gitattributes:2]: Keeping `bat` files Windows compatible is probably good idea

**Review by owaiskazi19 — APPROVED** [review]
Based on https://github.com/diffplug/spotless/issues/39. This LGTM.

**Comment by owaiskazi19** [comment]
> On Linux:
pull
`./gradlew.bat` :benchmarks:spotlessJavaCheck - FAILURE`

I am still not sure why are we running `gradlew.bat` on Linux and not simply `./gradlew`. As clearly `.bat` is for Windows.

**Comment by reta** [comment]
> I am still not sure why are we running gradlew.bat on Linux and not simply ./gradlew. As clearly .bat is for Windows.

@owaiskazi19 `./gradlew.bat` as obviously a mistake (I copy/pasted the details from 2 different boxes), for Linux, should be `./gradlew`, the outcome is the same.

> The way I was thinking for #1637 is to run SpotlessApply and push those changes. That'll have a successful SpotlessJavaCheck.

Please go ahead, as I explained to, I did that and I have a branch with those changes pushed, https://github.com/reta/OpenSearch/tree/issue-1637.apply, you are welcome to pull it and run `./gradlew :benchmarks:spotlessJavaCheck` on Linux, unless you have Git configuration in place to manipulate EOL (see please https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings), the build fails.

**Comment by dblock** [comment]
> The way I was thinking for #1637 is to run SpotlessApply and push those changes. That'll have a successful SpotlessJavaCheck.

Will it or will it start breaking on *nix? If yes, let's do that instead of modifying .gitattributes?

**Comment by reta** [comment]
> Will it or will it start breaking on *nix? If yes, let's do that instead of modifying .gitattributes? I am honestly not sure what "the right thing to do" is.

@dblock it fails for me

**Comment by dblock** [comment]
> > Will it or will it start breaking on *nix? If yes, let's do that instead of modifying .gitattributes? I am honestly not sure what "the right thing to do" is.
> 
> @dblock it fails for me on *nix, passes on Windows.

That's what I thought. Spotless needs to be told to be LF on Windows, which is done in this PR. I'm going to merge this.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
@owaiskazi19 Did I miss anything? Merge?

**Comment by reta** [comment]
@dblock fyi, just to point out a few references, all related to `spotless`:
 - https://github.com/diffplug/spotless/blob/main/.gitattributes
 - https://github.com/junit-team/junit5/blob/main/.gitattributes
 - ... (many more)

**Comment by owaiskazi19** [comment]
> @owaiskazi19 Did I miss anything? Merge?

Since Spotless copies git's behavior. This change will set Spotless to use LF and that's it. I don't see any issue in merging it. Let's see what it open for us in the future. 
Thanks @reta for the efforts and getting this change in!

**Comment by reta** [comment]
Thanks @owaiskazi19 !

**Comment by reta** [comment]
@dblock sorry, I am a bit confused by "backport" process, is it automated or still manual? (I don't see backport pull request to 1.x)

**Comment by VachaShah** [comment]
Hi @reta, it is automated but the label `backport 1.x` has to be added before the PR is merged and the workflow should run as part of the checks on the PR which doesn't look to be the case with this PR. I am not sure if there is a way to re-run the checks for a merged PR.

This is an example of how it should run https://github.com/VachaShah/TestGithubActions/pull/6 (its a test repo I have) here I added the `backport release-branch` label and the Backport workflow ran as a check on the PR. Same thing should happen on this PR for it to detect that a backport is requested.

**Comment by reta** [comment]
@VachaShah ah ... this is what I suspected, thank you, AFAIK external contributors do not have permissions to add labels (at least, that was certainly true before not long ago). I will try to add one next time, thank you!

**Comment by dblock** [comment]
> Hi @reta, it is automated but the label `backport 1.x` has to be added before the PR is merged 

Oh! I didn't know that. So if I see a PR that I want backported, I can't just label it and then merge? What should I do in that case?

**Comment by VachaShah** [comment]
> > Hi @reta, it is automated but the label `backport 1.x` has to be added before the PR is merged
> 
> Oh! I didn't know that. So if I see a PR that I want backported, I can't just label it and then merge? What should I do in that case?

Yes you can add the label before the merge and wait for the workflow to run. The workflow will start running as soon as the label is added and finish in some seconds (I tried this on one of the PRs). Once it finishes, the PR can be merged and a backport PR will be opened automatically. For the previous PRs, the workflow was not running as the Github action was not added in the list that is allowed to run. It works now successfully.

**Pull request merged**
